text/0715-construct-upgrade-support.md

```diff
@@ -0,0 +1,1274 @@
+users with clearer, more actionable error messages.
+
+##### Change Set Analysis for `Retain-Remove-Import Migration`
+
+For the `Retain-Remove-Import Migration` strategy, we will introduce an additional option `--import` in the
+`cdk diff` command like `cdk diff --import` which will create a change set on the existing stack with
+`import` change set type. Note that auxiliary resources—such as IAM roles, policies, or custom resources may be added
+or removed as part of construct upgrade for this strategy, and will omitted during the validations. An example is as
+follows:
+
+```json
+$ cdk diff --import
+[
+  {
+    "type": "Resource",
+    "resourceChange": {
+      "action": "Import", // NOTE THAT THIS SHOWS "IMPORT"
+      "logicalResourceId": "MyTable794EDED1",
+      "physicalResourceId": "DemoStack-MyTable794EDED1-11W4MR8VZ0UPE",
+      "resourceType": "AWS::DynamoDB::GlobalTable",
+      "replacement": "True",
+      "scope": [],
+      "details": [],
+      "afterContext": "..."
+    }
+  },
+  {
+    "type": "Resource",
+    "resourceChange": {
+      "policyAction": "Retain",
+      "action": "Remove",
+      "logicalResourceId": "MyTable794EDED1",
+      "physicalResourceId": "DemoStack-MyTable794EDED1-11W4MR8VZ0UPE",
+      "resourceType": "AWS::DynamoDB::Table",
+      "scope": [],
+      "details": [],
+      "beforeContext": "..."
+    }
+  }
+]
+```
+
+The `construct-upgrade` CLI command will implicitly call `cdk diff --import` which will create an `import`
+type change set as above and it will validate if the main resources' action is `Import` instead of `Add`. Similar
+to above, the main resources can be determined based on the `--target` value.
+
+There may be future enhancement that we add to the change set validation, please see
+[Appendix D: Change Set Analysis Future Enhancements](#appendix-d-change-set-analysis-future-enhancements).
+
+#### [In-Place Migration Only] CloudFormation stack refactoring Validation
+
+For in-place migrations, the construct-upgrade CLI command ensures the correctness of logical ID mappings
+defined in `refactor.json` by validating against the deployed V1 template and synthesized V2 template.
+
+```json
+[
+    {
+        "Source": {
+            "StackName": "<stack-name>",
+            "LogicalResourceId": "vpcv1A85508D7"
+        },
+        "Destination": {
+            "StackName": "<stack-name>",
+            "LogicalResourceId": "vpcA2121C38"
+        }
+    },
+    {
+        "Source": {
+            "StackName": "<stack-name>",
+            "LogicalResourceId": "vpcv1ipv6cidr87175A16"
+        },
+        "Destination": {
+            "StackName": "<stack-name>",
+            "LogicalResourceId": "vpcamazonProvided373FDA5C"
+        }
+    },
+    ......
+]
+```
+
+The validation begins by verifying that every source LogicalId in `refactor.json` corresponds to a resource
+in the deployed V1 template, while every target LogicalId maps to a resource in the synthesized V2 template.
```

Comment:
What about mappings that the user may have forgotten? For example, the app has 2 subnets, but the `refactor.json` file only mentions 1 subnet. Will this be blocked by CFN?